### PR TITLE
Support iree-run-mlir to read input values from a file.

### PR DESCRIPTION
**Pull request by hanhanW**
part of https://github.com/google/iree/issues/2304

### CODE REVIEW

File: iree/tools/run_mlir_main.cc [hunk 1]
```diff
@@ -258,6 +265,22 @@ StatusOr<std::string> PrepareModule(
   return binary_contents;
 }
 
+// Returns a splitted input values from `filename` using comma as separater.
+std::vector<std::string> GetInputValues(const std::string& filename) {
+  std::string error_message;
+  auto file = mlir::openInputFile(filename, &error_message);
+  if (!file) {
+    NotFoundErrorBuilder(IREE_LOC)
```
Comment (benvanik) [inline iree/tools/run_mlir_main.cc:273]: Make this file return StatusOr and return this instead

Comment (hanhanW) [reply iree/tools/run_mlir_main.cc]: Done

File: iree/tools/run_mlir_main.cc [hunk 2]
```diff
@@ -258,6 +265,21 @@ StatusOr<std::string> PrepareModule(
   return binary_contents;
 }
 
+// Returns a splitted input values from `filename` using comma as separater.
+StatusOr<std::vector<std::string>> GetInputValues(const std::string& filename) {
+  std::string error_message;
+  auto file = mlir::openInputFile(filename, &error_message);
+  if (!file) {
+    return NotFoundErrorBuilder(IREE_LOC)
+           << "Unable to open input file " << filename << ": " << error_message;
```
Comment (GMNGeoffrey) [inline iree/tools/run_mlir_main.cc:274]: Nit: put filename in quotes

```suggestion
           << "Unable to open input file '" << filename << "': " << error_message;
```

File: iree/tools/run_mlir_main.cc [hunk 3]
```diff
@@ -258,6 +265,21 @@ StatusOr<std::string> PrepareModule(
   return binary_contents;
 }
 
+// Returns a splitted input values from `filename` using comma as separater.
```
Comment (GMNGeoffrey) [inline iree/tools/run_mlir_main.cc:268]: If we're going to be reading from a file anyway, it seems like having these be newline-separated would be better. That would also allow commas in the input tensors themselves

File: iree/tools/run_mlir_main.cc [hunk 4]
```diff
@@ -267,11 +289,20 @@ Status EvaluateFunction(iree_vm_context_t* context,
 
   std::cout << "EXEC @" << export_name << std::endl;
   ASSIGN_OR_RETURN(auto input_descs, ParseInputSignature(function));
-  auto input_values_list = absl::MakeConstSpan(
-      input_values_flag.empty() ? nullptr : &input_values_flag.front(),
-      input_values_flag.size());
-  ASSIGN_OR_RETURN(auto* input_list, ParseToVariantList(input_descs, allocator,
-                                                        input_values_list));
+  iree_vm_variant_list_t* input_list;
+  if (!input_values_file_flag.empty()) {
```
Comment (GMNGeoffrey) [inline iree/tools/run_mlir_main.cc:293]: Validation that only one of input_values_file_flag and input_values_flag is set?

File: iree/tools/run_mlir_main.cc [hunk 5]
```diff
@@ -267,11 +289,20 @@ Status EvaluateFunction(iree_vm_context_t* context,
...
+    ASSIGN_OR_RETURN(auto input_values, GetInputValues(input_values_file_flag));
+    ASSIGN_OR_RETURN(input_list, ParseToVariantList(
+                                     input_descs, allocator,
+                                     absl::MakeConstSpan(&input_values.front(),
+                                                         input_values.size())));
```
Comment (GMNGeoffrey) [inline iree/tools/run_mlir_main.cc:298]: Seems like explicit span creation here shouldn't be necessary. There should be an implicit conversion from vector to const span

**Comment by hanhanW** [comment]
> Would love to see this for iree-run-module and iree-benchmark-module also!

I plan to do it in a later PR!

File: iree/tools/run_mlir_main.cc (hunk 5 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc]: Oh I didn't find it, thanks!

File: iree/tools/run_mlir_main.cc (hunk 4 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc:294]: Done

File: iree/tools/run_mlir_main.cc (hunk 3 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc]: Yes, it's better. Let's do it. I was thinking to follow the format in iree-run-module.

btw, I'm not sure if it ("\n") is working on Windows... (we could worry about it later?)

File: iree/tools/run_mlir_main.cc (hunk 2 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc]: Done

File: iree/tools/run_mlir_main.cc [hunk 6]
```diff
@@ -258,6 +265,22 @@ StatusOr<std::string> PrepareModule(
...
+  std::string error_message;
+  auto file = mlir::openInputFile(filename, &error_message);
+  if (!file) {
+    return NotFoundErrorBuilder(IREE_LOC) << "Unable to open input file '"
+                                          << filename << "': " << error_message;
+  }
+  llvm::SmallVector<llvm::StringRef, 8> source_buffers;
+  file->getBuffer().split(source_buffers, /*Separator=*/"\n", /*MaxSplit=*/-1,
+                          /*KeepEmpty=*/false);
+  std::vector<std::string> res;
```
Comment (antiagainst) [inline iree/tools/run_mlir_main.cc:279]: res.reserve(source_buffers.size())?

File: iree/tools/run_mlir_main.cc [hunk 7]
```diff
@@ -122,6 +122,13 @@ static llvm::cl::list<std::string> input_values_flag{
     llvm::cl::ZeroOrMore,
 };
 
+static llvm::cl::opt<std::string> input_values_file_flag{
+    "input-value-file",
+    llvm::cl::desc(
+        "<input value file>, see run_module_main.cc for input format"),
```
Comment (antiagainst) [inline iree/tools/run_mlir_main.cc:128]: "Provides a file for input shapes and optional values (see ... for details)" ?

Actually I think we might what to put the format spec in https://google.github.io/iree/DeveloperOverview so we can point to a URL here. Source code is not always there. :)

**Comment by GMNGeoffrey** [comment]
FYI I'm going to add an android presubmit and I think this PR is old enough that it won't have the build config and so will fail. You can ignore an android failure you see on this, or rebase to pick up the config.

File: iree/tools/run_mlir_main.cc (hunk 6 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc:279]: Done

File: iree/tools/run_mlir_main.cc (hunk 7 above)
Comment (hanhanW) [reply iree/tools/run_mlir_main.cc]: Cool, I'll follow on this once all the tools get support of input file.

**Comment by hanhanW** [comment]
> FYI I'm going to add an android presubmit and I think this PR is old enough that it won't have the build config and so will fail. You can ignore an android failure you see on this, or rebase to pick up the config.

I tried to rebase but it asked me to fix more and more files as expected, like bunch of CMakelists.txt... so I just ignore the failure and merge it. :/

**Pull request merged**